src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp

```diff
@@ -93,46 +128,77 @@ void Bank::evaluate_and_allocate() {
-            // Evaluate concurrently, lock the device
-            // mutex only to update the device bank (& allocate on-device memory, if needed)
-            const auto& transformed_tensor = lt.eval();
+            const auto& transformed_tensor_meta = lt.eval_meta();
 
             std::unique_lock<std::mutex> guard(device_bank.mutex);
+            uids_to_meta[uid] = transformed_tensor_meta;
             if (device_for_alloc == "CPU") {
                 // No allocation needed
-                device_bank.storage[device_bank.registered_tensors.at(lt)].tensor = transformed_tensor;
+                guard.unlock();
+                auto transformed = lt.eval();
+                std::unique_lock guard(device_bank.mutex);
+                device_bank.storage.at(uid).tensor = std::move(transformed);
                 return;
             }
 
+            // Need to allocate
+            // Note: secured by the lock above
+            uids_to_allocated.insert({uid, ov::Tensor()});
+        });
+
+        std::unique_lock guard(device_bank.mutex);
+        // Allocate memory sequentially - in order of UID
+        for (const auto& elem : uids_to_allocated) {
+            auto uid = elem.first;
+
             ov::SoPtr<ov::ITensor> remote_tensor;
             ov::Tensor allocated_tensor;
 
-            auto remote_ctx = m_core->get_default_context(device_for_alloc)._ptr;
-            remote_tensor =
-                remote_ctx->create_host_tensor(transformed_tensor.get_element_type(), transformed_tensor.get_shape());
-            allocated_tensor = ov::make_tensor(remote_tensor);
-            device_bank.storage[device_bank.registered_tensors.at(lt)].tensor = allocated_tensor;
-            guard.unlock();  // Unlock the guard, map update is done - copy can continue in parallel
+            remote_tensor = ov::npuw::weights::context(m_core, device_for_alloc)
+                                ->create_host_tensor(uids_to_meta[uid].type, uids_to_meta[uid].shape);
+            uids_to_allocated.at(uid) = ov::make_tensor(remote_tensor);
+        }
+        guard.unlock();
 
-            transformed_tensor.copy_to(allocated_tensor);
+        // Copy to allocated memory
+        // Evaluate concurrently, lock the device
+        // mutex only to update the device bank (& allocate on-device memory, if needed)
+        ov::parallel_for(uids_to_allocated.size(), [&](std::size_t idx) {
+            std::unique_lock dev_guard(device_bank.mutex);
+            auto it = uids_to_allocated.begin();
+            // FIXME: linear complexity
+            std::advance(it, idx);
+            auto uid = it->first;
+            auto allocated_tensor = it->second;
+            dev_guard.unlock();
+
+            auto transformed = device_bank.storage.at(uid).lt.eval();
+            transformed.copy_to(allocated_tensor);
+
+            std::unique_lock<std::mutex> guard(device_bank.mutex);
+            device_bank.storage.at(uid).tensor = std::move(allocated_tensor);
```

Comment:
it seems you don't need this lock, who else would write at the same location there? and since it is `.at()`, a new element is not created in the shared (presumably, concurrently accessed) map?

Reply:
Done